server/resp/src/main/java/org/infinispan/server/resp/Resp3Handler.java

```diff
@@ -49,11 +48,18 @@ public class Resp3Handler extends Resp3AuthHandler {
    public void setCache(AdvancedCache<byte[], byte[]> cache) {
       super.setCache(cache);
       ignorePreviousValueCache = cache.withFlags(Flag.SKIP_CACHE_LOAD, Flag.IGNORE_RETURN_VALUES);
-      Cache toMultimap = cache.withMediaType(MediaType.APPLICATION_OCTET_STREAM, valueMediaType);
-      listMultimap = new EmbeddedMultimapListCache<>(toMultimap);
-      mapMultimap = new EmbeddedMultimapPairCache<>(toMultimap);
-      embeddedSetCache = new EmbeddedSetCache<>(toMultimap);
-      sortedSetMultimap = new EmbeddedMultimapSortedSetCache<>(toMultimap);
+      listMultimap = new EmbeddedMultimapListCache<>(getCache(cache));
+      mapMultimap = new EmbeddedMultimapPairCache<>(getCache(cache));
+      embeddedSetCache = new EmbeddedSetCache<>(getCache(cache));
+      sortedSetMultimap = new EmbeddedMultimapSortedSetCache<>(getCache(cache));
+   }
+
+   @SuppressWarnings({"unchecked", "rawtypes"})
+   private <K, V> Cache<K, V> getCache(AdvancedCache cache) {
```

Comment:
I'm not sure that I understand the use of this method. Can't `setCache` use `typedCache` directly?

Reply:
Was double-checking this one. We need it to create the proper decoration. With this method, if the cache is DecoratedCache (after an EXEC), we call the `withMediaType` and create `DecoderCache` -> `DecoratedCache` -> `CacheImpl/SimpleCacheImpl` chain.

If we call just the `typedCache` method, it might return the `DecoratedCache` -> `CacheImpl/SimpleCacheImpl` chain.

I'll add this as a comment, otherwise, I forget what it does by tomorrow.

Reply:
Where does the `DecoderCache` come from? I assume it is `EncoderCache` (there is the same typo in the comment).
What is the problem with `DecoratedCache` -> `[Simple]CacheImpl`? The `typedCache` method invokes only when it needs to re-encode the data as far as I understand. What am I missing?

Reply:
Actually, I'll need to revert this whole block, it should always call the withMediaType before passing the instance to the multimap implementations. I'll add it as a comment, though.

Basically, the functional map implementation has a conversion in place (e.g., byte[] -> WrappedByteArray). If we don't pass an `EncoderCache` to these multimap implementations, they would try accessing entries with key (byte[]) instead for some operations. I thought we could use the `typedCache` to avoid that call in some cases, but we can not. Thanks for asking!